smack-core/src/main/java/org/jivesoftware/smack/AbstractXMPPConnection.java

```diff
@@ -2012,7 +2012,7 @@ protected void sendStreamOpen() throws NotConnectedException, InterruptedExcepti
             from = XmppStringUtils.completeJidFrom(localpart, to);
         }
         String id = getStreamId();
-        sendNonza(new StreamOpen(to, from, id));
+        sendNonza(new StreamOpen(to, from, id, config.getLanguage().toString().replace("_", "-"), StreamOpen.StreamContentNamespace.client));
```

Comment:
This could cause an NPE if `language` is `null`. I suggest to create a `ConnectionConfiguration.getLanguageTag()` method which returns `null` if language is `null`. This is also the method the TODO comment should go to, which I am still missing.

Reply:
Oh, right. The TODO comment! Sorry about that! Is there a particular reason why `getLanguageTag` should return `null` and not the default "en" value?

Reply:
> Sorry about that! Is there a particular reason why `getLanguageTag` should return `null`

It is not `getLanguageTag()` that I am worried about returning `null`, it is `getLanguage()`. Currently it is possible to set the language to `null` in the configuration, and that is desirable given that `xml:lang` in stream open elements is optional. But the code should take that into consideration and not throw NPEs in that, arguably rare, case.

Reply:
Understood. The questions that I'm looking at, are:
- is it reasonable to allow the user to supply a null Locale?
- what is the expected outcome in that case? "en", "" or no xml:lang tag at all?

Reply:
> * is it reasonable to allow the user to supply a null Locale?

I think so, especially for integration tests and the like. The point is, the standard is designed with an non-existing xml:lang in the stream open element in mind, and a library implementing the standard should allow for it. Of course, users deliberately providing a `null` value here, go down the "you should know what you are doing" road. And that is why Smack defaults to a non-empty xml:lang.

> * what is the expected outcome in that case? "en", "" or no xml:lang tag at all?

No xml:lang at all, otherwise the user would have provided `Locale.ENGLISH` for "en" or a `Locale` which returns the empty string as language tag for "".

The latter case is especially interesting: values where the empty string is allowed often cause pain, as, for example, their semantic is usually not as clearly defined as people believe to be or cause different behavior by different implementations. I really would like to be able to investigate how server implementations behave if a stream open with `xml:lang=""` is send (AFAIKT it is not explicitly forbidden by the standard). That said, given the current design we are aiming for, it is not possible to configure Smack to do so. That is because

```java
jshell> var l = new java.util.Locale("");
l ==> 
jshell> l.getLanguage()
$3 ==> ""
jshell> l.toLanguageTag()
$4 ==> "und"
```

so once we use `toLangaugeTag()` which is the goal, using a Locale constructed with the empty string as sole argument, will cause toLanguageTag() to return "und" (which I assume stands for 'undefined'). Hence the stream open will carry `xml:lang="und".

That means we should probably not introduce a method `ConnectionConfiguration.getLanguageTag()`, but `ConnectionConfiguration.getXmlLang()`, which
- returns `null` if `language` is null
- returns "" if `language.getLanguage()` returns the empty string
- returns `language.toLanguageTag()` otherwise